### DISCUSSION TITLE
Multi-repo Agent

### DISCUSSION THREAD

**Discussion by BradKML** [SmartManoj/Kevin#137] [Ideas]
Here is a small thought: imaging if the agent can
- enhance "core" repos using other repo(s) as inspiration or example implementation
- use old repo as reference to make new repo more feature-complete (with less reliance on documentation)

**Comment by SmartManoj** [comment]
CodeAct Agent itself will do this.

**Reply by BradKML** [reply]
Q: sometimes OpenHands will crawl into package repos and fix them, but other times it will not do so, how can CodeAct Agent know the process of improving a package when it is not in the working file path, but rather the venv path? Also, how hard is it to tell an AI to not touch the reference libraries?

**Reply by SmartManoj** [reply]
> but other times it will not do so,

you can explicitly prompt it if you want it to do like that.

**Reply by BradKML** [reply]
Probably gonna be slightly redundant here, but there has to be sample prompts for OpenHands with this important use case.

**Reply by SmartManoj** [reply]
> but other times it will not do so,

Could you share an MRE?

**Reply by BradKML** [reply]
As a kind of example (just to see how it operates, not to yield a final product) cus this seems a bit more challenging

Dynamic 1: library size
- **Borrow big library features to small libraries:** Given the diverse models and HPOs of PyCaret, see what can be copied into a small library like LazyPredict regarding regression (or classification, or HPO enhancements) since between they do not match completely https://pycaret.readthedocs.io/en/stable/api/regression.html https://github.com/nityansuman/lazypredict-nightly?tab=readme-ov-file#regression
- **Integrating small library features to large libraries:** Given new algorithms like Perpetual and ChefBoost, integrate that into core repos like Scikit-Learn https://github.com/serengil/chefboost https://github.com/perpetual-ml/perpetual

Dynamic 2: staleness or general usage
- **Extract stale library features to popular libraries** Given a stale library (Backtrader), copy the core unique features to the current best (backtesting.py) https://github.com/backtrader2/backtrader https://github.com/kernc/backtesting.py
- **Move SOTA library features to mainstream libraries** Given Orion has some of the newer algorithms like DEHB, port that back into Ray-Tune which more people will use it https://orion.readthedocs.io/en/stable/user/algorithms.html https://docs.ray.io/en/latest/tune/api/suggestion.html 

Checking criteria:
1. The agents should not be prompted which specific algorithm exists in one place and not the other
2. The agents are allowed to search for "readthedocs" or other documentation (Repo or Online)
3. Prompt can be customized such that specific features are the main goal
4. Prompt can be generalized to using multiple repos as sources for one destination repo
5. Pull Request sizing recommendation applies (smaller and more modular is better)
